modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -480,6 +625,152 @@ public int getZoneId(String name) {
         }
     }
 
+    /**
+     * The method for obtaining data nodes of the specified zone. If the dataNodesAutoAdjustScaleUp or dataNodesAutoAdjustScaleDown
+     * equals to 0 then returned completable future will be completed when data nodes will be saved to the meta storage.
+     *
+     * @param zoneId Zone id.
+     * @param topVer Topology version.
+     * @return The data nodes future.
+     */
+    public CompletableFuture<Set<String>> getDataNodes(int zoneId, long topVer) {
```

Comment:
Please, divide this method and extract part when we react on future completion, it will increase readability.  Also please add comments right inside the code, currently it's auite hard to understand what's going on

Reply:
I rework this code and add comments.